shared_model/utils/lazy_initializer.hpp

```diff
@@ -29,47 +29,42 @@ namespace shared_model {
      * @tparam Source - input type
      * @tparam Target - output type
      */
-    template <typename Source, typename Target>
+    template <typename Target>
     class LazyInitializer {
      private:
-      /// Type of transformation
-      using TransformType = std::function<Target(const Source &)>;
+      /// Type of generator function
+      using GeneratorType = std::function<Target()>;
 
      public:
-      LazyInitializer(const Source &source, const TransformType &transform)
-          : source_(source), transform_(transform) {}
+      LazyInitializer(const GeneratorType &generator) : generator_(generator) {}
 
       LazyInitializer(const LazyInitializer &) = default;
 
       /**
-       * @return value after transformation
+       * @return generated value
        */
-      const Target &get() {
+      const Target &get() const {
         if (target_value_ == nullptr) {
-          target_value_ = std::make_shared<Target>(transform_(source_));
+          target_value_ = std::make_shared<Target>(generator_());
         }
         return *target_value_;
       }
 
      private:
-      const Source &source_;
-      TransformType transform_;
-      std::shared_ptr<Target> target_value_;
+      GeneratorType generator_;
+      mutable std::shared_ptr<Target> target_value_;
     };
 
     /**
      * Function for creating lazy object
-     * @tparam Source - source type
-     * @tparam Transform - type of transformation
-     * @param source - source value
-     * @param transform - transformation instance
+     * @tparam Generator - type of generator
+     * @param generator - instance of Generator
      * @return initialized lazy value
      */
-    template <typename Source, typename Transform>
-    auto makeLazyInitializer(Source &&source, Transform &&transform) {
-      using targetType = decltype(transform(source));
-      return LazyInitializer<Source, targetType>(std::forward<Source>(source),
-                                      std::forward<Transform>(transform));
+    template <typename Generator>
+    auto makeLazyInitializer(Generator &&generator) {
+      using targetType = decltype(generator());
```

Comment:
Too generic for me, maybe restrict it's declaration with smth like:
```cpp
static_assert(!std::is_lvalue_reference<targetType>::value)
```

Reply:
There are no reasons to make static assert, but we need to deduce the type of `Target`. We know that generator is a functor, that return `Target` type.

Reply:
If `Target` is _lvalue ref_ than it's something wrong here

Reply:
Well, we may handle reference in `shared_ptr`, isn't it? But usage defined by the user if he or she wants to store reference, we are not allowed to deny it.

Reply:
Let it be, just imo that can reduce this types of errors